From 04a58ea7ead4f99d451425a8dfaa0c13bb16062e Mon Sep 17 00:00:00 2001
From: Johannes Schindelin <johannes.schindelin@gmx.de>
Date: Thu, 4 Jul 2019 14:25:49 +0200
Subject: [PATCH 5/5] Fix INTERACTIVE(fd) on Windows

On Windows, `isatty()` also reports success on character devices such as
`NUL`, even if it is not an interactive console. That is, however, what
PCRE2 expects `isatty()` to determine. Let's work around that.

Signed-off-by: Johannes Schindelin <johannes.schindelin@gmx.de>
---
 src/pcre2test.c | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/src/pcre2test.c b/src/pcre2test.c
index ea52a20..0f62ecd 100644
--- a/src/pcre2test.c
+++ b/src/pcre2test.c
@@ -106,6 +106,30 @@ this needs to be updated or the build could break */
 /* Put the test for interactive input into a macro so that it can be changed if
 required for different environments. */
 
+#if defined(_WIN32) || defined(WIN32)
+static int win32_isatty(int fd)
+{
+	HANDLE h;
+	DWORD mode;
+	CONSOLE_SCREEN_BUFFER_INFO info;
+
+	if (!isatty(fd))
+		return 0;
+
+	h = (HANDLE) _get_osfhandle(fd);
+	if (GetFileType(h) != FILE_TYPE_CHAR)
+		return 0;
+	if (!fd) {
+		if (!GetConsoleMode(h, &mode))
+			return 0;
+	} else if (!GetConsoleScreenBufferInfo(h, &info))
+		return 0;
+
+	return 1;
+}
+#define isatty win32_isatty
+#endif
+
 #define INTERACTIVE(f) isatty(fileno(f))
 
 
-- 
2.36.1.windows.1

